python/ray/data/dataset.py

```diff
@@ -4047,52 +4056,53 @@ def __iter__(self):
         pipe = DatasetPipeline(it, False, length=len(it._splits))
         if read_stage:
             pipe = pipe.foreach_window(
-                lambda ds, read_stage=read_stage: Dataset(
+                lambda ds, read_stage=read_stage: Datastream(
                     ds._plan.with_stage(read_stage), ds._epoch, True
                 )
             )
         return pipe
 
-    @Deprecated(message="Use `Dataset.cache()` instead.")
-    def fully_executed(self) -> "Dataset[T]":
+    @Deprecated(message="Use `Datastream.cache()` instead.")
+    def fully_executed(self) -> "MaterializedDatastream[T]":
```

Comment:
after this PR, `ds.fully_executed()/cache()` will have behavior change, that `ds.fully_executed()/cache()` will trigger execution each time of call. Before this PR, only first call of `ds.fully_executed()/cache()` triggers execution.

Do we want to document the behavior change somewhere in doc?

Reply:
I added a bit of explanation to the pydoc here, but I don't believe we need to further justify the previous behavior since that was never well defined / documented (in fact it seems like a bug).

Reply:
Ok, I was worried about people depending on the behavior. Cool, let's start with this.

Reply:
It seemed like this did affect one existing code path--- one test with DatasetPipeline with an optimization flag disabled. Since this path is not documented, I opted to remove it. We could also roll back the behavior (have the mutating behavior) as an alternative, and delay changing this until we remove DatasetPipeline. However, I think the likelihood someone is depending on this flag is low.

Reply:
It's fine with me to change the behavior here.